clippy_lints/src/rc_clone_in_vec_init.rs

```diff
@@ -47,27 +49,107 @@ declare_lint_pass!(RcCloneInVecInit => [RC_CLONE_IN_VEC_INIT]);
 impl LateLintPass<'_> for RcCloneInVecInit {
     fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
         let Some(macro_call) = root_macro_call_first_node(cx, expr) else { return; };
-        let Some(VecArgs::Repeat(elem, _)) = VecArgs::hir(cx, expr) else { return; };
+        let Some(VecArgs::Repeat(elem, len)) = VecArgs::hir(cx, expr) else { return; };
         let Some(symbol) = new_reference_call(cx, elem) else { return; };
+        let lint_span = macro_call.span;
+        let symbol_name = symbol.as_str();
+        let len_snippet = snippet(cx, len.span, "..");
+        let elem_snippet = elem_snippet(cx, elem, symbol_name);
+        let indentation = indent_of(cx, lint_span).unwrap_or(0);
+        let lint_suggestions =
+            construct_lint_suggestions(lint_span, symbol_name, &elem_snippet, len_snippet.as_ref(), indentation);
 
-        emit_lint(cx, symbol, &macro_call);
+        emit_lint(cx, symbol, lint_span, &lint_suggestions);
     }
 }
 
-fn emit_lint(cx: &LateContext<'_>, symbol: Symbol, macro_call: &MacroCall) {
+struct LintSuggestion {
```

Comment:
I;m wondering whether this can be extracted to a util somewhere

Reply:
I think in most cases it's better to use `span_lint_and_then` and pass the `DiagnosticBuilder` around. That's also a bit more efficient when it comes to performance. The `DiagnosticBuilder` will store the suggestions internally. Is there a reason why you want to use this intermediate struct to store the suggestions? :upside_down_face:

Reply:
Just so I can construct the suggestions separately from where I'm emitting the lint, so that the `span_lint_and_then` code will be clean from logistics.

In terms of performance, since I'm passing a `& mut Diagnostic` the performance overhead is basically only the `Vec<LintSuggestion>` creation - which I thought is a small price to pay for readability.

Lmk if there's some other design you'd prefer 🙂

Reply:
The idea in rustc is actually to create the suggestion inside the closure. Clippy does it a bit differently to have a simpler interface, but if we use `span_lint_and_then` we usually try to also create the suggestion inside. Extracting the specific string creation to single functions like this is nice, but I wouldn't store them in another struct. Also, since the `Span` and `Applicability` is always the same.

> In terms of performance

This comment was more about the internal implementation in rustc. If lints are allowed, they won't be emitted, rustc then never calls the closure. That's what I was referring to. That is super nit picky though, it only came to mind since I've worked on that recently. I sometimes go a bit overboard with my reviews, this side node is an example for this :sweat_smile:. So please ignore the performance comment :upside_down_face:

Reply:
I'm willing to go as deep into this rabbit hole as you are 🙂 
np, I'll change the design a bit so that everything is being passed to `emit_lint` and it'll construct the suggestion by calling another function inside the clousure

Reply:
Alright, now the question remains, what is the benefit of this struct. I think that adding everything together in the `span_lint_and_then` closure would look a bit cleaner (and avoid a vec allocation). For example, it would look like this:

```rs
    span_lint_and_then(
        cx,
        RC_CLONE_IN_VEC_INIT,
        lint_span,
        &format!("calling `{symbol_name}::new` in `vec![elem; len]`"),
        |diag| {
            let len_snippet = snippet(cx, len.span, "..");
            let elem_snippet = elem_snippet(cx, elem, symbol_name);
            let indentation = indent_of(cx, span).unwrap_or(0);
            let indentation = " ".repeat(indentation);
            let suggestions = construct_lint_suggestions(cx, lint_span, symbol_name, elem, len);

            diag.note(format!("each element will point to the same `{symbol_name}` instance"));
            diag.span_suggestion(
                lint_span,
                &format!("consider initializing each `{symbol_name}` element individually"),
                &loop_init_suggestion(&elem_snippet, len_snippet.as_ref(), &indentation),
                Applicability::Unspecified,
            );
            diag.span_suggestion(
                lint_span,
                &format!("or if this is intentional, consider extracting the `{symbol_name}` initialization to a variable"),
                &extract_suggestion(&elem_snippet, len_snippet.as_ref(), &indentation),
                Applicability::Unspecified,
            );
        },
    );
```

This way everything is together in one place, and it still seems very readable to be :upside_down_face:

> I'm willing to go as deep into this rabbit hole as you are 

Okay, how much time do you have? :joy: JK, I only looked at it in that detail to implement something else in rustc

Reply:
Didn't notice your comment there, oops 😄 

I really want the closure to be as clean as possible, this is just my style. I ended up still having a struct and a vec, but now everything is being called from within the closure.
Its somewhere in the middle between your idea and mine haha

If you really want to get rid of the struct and the vec, I can change `construct_lint_suggestions` to receive a `&mut Diagnostic` as well, and it'll call `diag.span_lint` inside.

I think its less readable to encapsulate the `diag` api calls inside `construct_lint_suggestions`, but i don't really mind if it'll sit better with you 🙂

Reply:
Question is, how important are performance here? Is vec allocation going to cost too much here?

Reply:
Not really, this is more about the style, I don't really see the reason for having the struct just to pass them to the `DiagnosticBuilder` afterwards. What about passing in the `DiagnosticBuilder` into the function as an argument?

Reply:
that's basically what I offered, just don't know what is `DiagnosticBuilder` exactly 🙈 
the closure receives a `&mut Diagnostic`.
I'm offering to pass it into `construct_lint_suggestions` such that this function will span the suggestions instead of returning them as a vec.

> I don't really see the reason for having the struct just to pass them to the DiagnosticBuilder afterwards.

Imo passing **named** data between scopes is arguably justifying a struct.
None the less, I'll change it as you wish of course 🙂

Reply:
I ended up basically taking your suggestion from here: https://github.com/rust-lang/rust-clippy/pull/8814#discussion_r870730625

Reply:
> Imo passing named data between scopes is arguably justifying a struct.

That's totally correct. In this case, I just felt a bit weird, since I would say that the `DiagnosticBuilder` is exactly for this. Sorry for not explaining what the builder is, it's just the argument received from the `span_lint_and_then` argument. This implementation looks good to me on a quick glance, I'll try to do a full review soon :upside_down_face: